### PR TITLE
IMPRO-1476: Accumulate CLI

### PR DESCRIPTION
**Pull request by neilCrosswaite**
# Accumulate CLI
This CLI takes the inputs of U and V and extrapolates at a fixed 1 min interval.

I have had to compress some .processes onto the delaration to make codacy happy. Although im not happy, so I may look into alternatives.

## Testing:
Tests have been written for accumulate as a bats test and for the new cli `create_constrained_inputcube_converter` function.
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/cli/nowcast_accumulate.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+from improver import cli
+
+inputadvection = cli.create_constrained_inputcube_converter(
+    'precipitation_advection_x_velocity', 'precipitation_advection_y_velocity')
+
+
+@cli.clizefy
+@cli.with_output
+def process(input_cube: cli.inputcube, advection_cubes: inputadvection,
```
Comment (benfitzpatrick) [inline lib/improver/cli/nowcast_accumulate.py]: Codacy Issue found: [Too many local variables (22/20)](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

**Comment by arh89** [comment]
@neilCrosswaite, there are a huge number of commits for this PR - could we possibly rebase and squash them all if they're no longer necessary please?

File: lib/improver/cli/__init__.py [hunk 2]
```diff
@@ -220,6 +220,36 @@ def with_intermediate_output(wrapped, *args, intermediate_output=None,
...
+
+        Args (iris.cube.Cube):
+            To cube to be passed forward for returning or loading.
+
+        Returns:
+            iris.cube.CubeList:
+                The loaded cubelist of constrained cubes.
+        """
+        from improver.utilities.load import load_cube
+        import iris
```
Comment (arh89) [inline lib/improver/cli/__init__.py]: If we only need `iris.cube.CubeList` , can we just `import` this please?

File: lib/improver/cli/__init__.py [hunk 3]
```diff
@@ -220,6 +220,36 @@ def with_intermediate_output(wrapped, *args, intermediate_output=None,
     return result
 
 
+def create_constrained_inputcube_converter(*constraints):
```
Comment (arh89) [inline lib/improver/cli/__init__.py]: Returns a `CubeList`, so can we rename this (and the inner function) as `inputcubelist` please?

File: lib/improver/cli/__init__.py [hunk 4]
```diff
@@ -220,6 +220,36 @@ def with_intermediate_output(wrapped, *args, intermediate_output=None,
...
+    Returns:
+        list of function:
+            A list of functions with different constraints.
+    """
+    @value_converter
+    def constrained_inputcube_converter(to_convert):
+        """Passes the cube and constraints onto maybe coerce with.
+
+        Args (iris.cube.Cube):
+            To cube to be passed forward for returning or loading.
```
Comment (arh89) [inline lib/improver/cli/__init__.py]: `to_convert`, as with the `inputjson` and `inputcube` converters is a string (filepath) (or cubelist)

File: lib/improver/cli/__init__.py (hunk 3 above)
Comment (arh89) [inline lib/improver/cli/__init__.py]: Also can we move this whole block of code next to the other converters, i.e: above the comment line `# output handling`

File: lib/improver/cli/nowcast_accumulate.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
```
Comment (arh89) [inline lib/improver/cli/nowcast_accumulate.py]: To match the other `clize` CLIs, can we separate out each of the parameters onto a new line please?

File: lib/improver/cli/nowcast_accumulate.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            e.g. 'mm'
+            Default is 'm'.
+
+    Returns:
+        iris.cube.CubeList:
+            New cubes with accumulated data.
+
+    Raises:
+        ValueError:
+            can either use speed_cube and direction_cube or u_cube and v_cube.
```
Comment (arh89) [inline lib/improver/cli/nowcast_accumulate.py]: Mistake here? Seems to raise if either u or v are missing.

File: lib/improver/cli/nowcast_accumulate.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    v_cube = advection_cubes.extract(
+        Constraint("precipitation_advection_y_velocity"), True)
+
+    if not (u_cube and v_cube):
+        raise TypeError(
+            "Neither u_cube or v_cube can be none")
+    oe_cube = merge_cubes(CubeList(oe_cube))
+
+    # The accumulation frequency in minutes.
+    accumulation_fidelity = 1
```
Comment (arh89) [inline lib/improver/cli/nowcast_accumulate.py]: Is there a reason for not having this as an optional argument that we pass in?

File: lib/improver/cli/nowcast_accumulate.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    'precipitation_advection_x_velocity', 'precipitation_advection_y_velocity')
+
+
+@cli.clizefy
+@cli.with_output
+def process(input_cube: cli.inputcube, advection_cubes: inputadvection,
+            *oe_cube: cli.inputcube, attributes_dict: cli.inputjson = None,
+            max_lead_time=360, lead_time_interval=15, accumulation_period=15,
+            accumulation_units='m'):
+    """Module to extrapolate and accumulate the weather with 1 min fidelity.
```
Comment (arh89) [inline lib/improver/cli/nowcast_accumulate.py]: Is there anything preventing us passing this in, so it's not just 1 minute?

File: lib/improver/tests/cli/test_init.py [hunk 9]
```diff
@@ -37,7 +37,8 @@
 from improver.cli import (
     docutilize, unbracket,
     maybe_coerce_with, inputcube, inputjson, with_output,
-    with_intermediate_output)
+    with_intermediate_output, create_constrained_inputcube_converter)
+from improver.utilities.load import load_cube
```
Comment (arh89) [inline lib/improver/tests/cli/test_init.py]: I don't think we should be loading any cubes here - so shouldn't be importing

File: lib/improver/tests/cli/test_init.py [hunk 10]
```diff
@@ -55,6 +56,7 @@ def dummy_function(first, second=0, third=2):
         (iris.cube.Cube)
 
     """
+
```
Comment (arh89) [inline lib/improver/tests/cli/test_init.py]: Unnecessary change? pycodestyle?

File: lib/improver/cli/__init__.py (hunk 2 above)
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: Done

File: lib/improver/cli/__init__.py (hunk 3 above)
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: Done

Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: Done

File: lib/improver/cli/nowcast_accumulate.py (hunk 5 above)
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_accumulate.py]: Done

File: lib/improver/cli/nowcast_accumulate.py (hunk 6 above)
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_accumulate.py]: `If advection_cubes doesn't contain an x velocity and a y velocity.`

File: lib/improver/cli/nowcast_accumulate.py (hunk 7 above)
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_accumulate.py]: It is meant to be fixed to 1 minute.

File: lib/improver/cli/__init__.py (hunk 4 above)
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: Done

File: lib/improver/cli/nowcast_accumulate.py (hunk 8 above)
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_accumulate.py]: Just following the ticket, open to discussion though.

File: lib/improver/tests/cli/test_init.py (hunk 10 above)
Comment (neilCrosswaite) [reply lib/improver/tests/cli/test_init.py]: Fixed

File: lib/improver/tests/cli/test_init.py [hunk 11]
```diff
@@ -185,6 +185,20 @@ def test_with_output(self, m):
...
+    """Tests the creature constraint inputcube converter"""
+
+    @patch('improver.cli.maybe_coerce_with', return_value='return')
+    def test_basic(self, m):
+        """Tests that it returns a function which itself returns 2 cubes"""
+        result = create_constrained_inputcubelist_converter(
+            'wind_speed', 'wind_from_direction')
+        result("foo")
+        m.assert_any_call(load_cube, "foo", constraints='wind_speed')
+        m.assert_any_call(load_cube, "foo", constraints='wind_from_direction')
```
Comment (benfitzpatrick) [inline lib/improver/tests/cli/test_init.py]: Codacy Issue found: [Undefined variable 'load_cube'](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

File: lib/improver/cli/nowcast_accumulate.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Script to accumulate input data given advection velocity fields."""
+
+from improver import cli
+
+inputadvection = cli.create_constrained_inputcube_converter(
```
Comment (benfitzpatrick) [inline lib/improver/cli/nowcast_accumulate.py]: Codacy Issue found: [Module 'improver.cli' has no 'create_constrained_inputcube_converter' member](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948)

File: lib/improver/tests/cli/test_init.py [hunk 13]
```diff
@@ -185,6 +185,20 @@ def test_with_output(self, m):
...
+class Test_create_constrained_inputcube_converter(unittest.TestCase):
+    """Tests the creature constraint inputcube converter"""
+
+    @patch('improver.cli.maybe_coerce_with', return_value='return')
+    def test_basic(self, m):
+        """Tests that it returns a function which itself returns 2 cubes"""
+        result = create_constrained_inputcubelist_converter(
+            'wind_speed', 'wind_from_direction')
+        result("foo")
+        m.assert_any_call(load_cube, "foo", constraints='wind_speed')
```
Comment (benfitzpatrick) [inline lib/improver/tests/cli/test_init.py]: Codacy Issue found: [Undefined variable 'load_cube'](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948)

File: lib/improver/tests/cli/test_init.py (hunk 9 above)
Comment (neilCrosswaite) [reply lib/improver/tests/cli/test_init.py]: In the mocking, it is testing that load_cube is called with the constraint.
```python
        m.assert_any_call(load_cube, "foo", constraints='wind_speed')
        m.assert_any_call(load_cube, "foo", constraints='wind_from_direction')
```

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

File: lib/improver/tests/cli/test_init.py (hunk 9 above)
Comment (arh89) [reply lib/improver/tests/cli/test_init.py]: Ah, my bad - if it's being mocked then I'm happy.

File: lib/improver/cli/nowcast_accumulate.py (hunk 8 above)
Comment (arh89) [reply lib/improver/cli/nowcast_accumulate.py]: @MoseleyS & @cgsandford - deciding vote is with you, as you know more about this than I do.

Comment (cgsandford) [reply lib/improver/cli/nowcast_accumulate.py]: We would like the accumulation fidelity always to be one minute.  This has been used before and is a sensible interval that is as fine as possible (to ensure we don't "skip" pixels between frames) without being ridiculous.  We do not anticipate the user changing this, and it is sensible for this to be consistent at a high level.  Hence it was suggested to fix this value in the code.

**Comment by neilCrosswaite** [comment]
I have also prepared pytests for this whenever it is needed.

File: lib/improver/cli/nowcast_accumulate.py (hunk 8 above)
Comment (arh89) [reply lib/improver/cli/nowcast_accumulate.py]: No problem with me then.

My only suggestion: if it's a constant, to change the variable name to all caps to make this clear.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

File: lib/improver/cli/nowcast_accumulate.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+            New cubes with accumulated data.
+
+    Raises:
+        TypeError:
+            If advection_cubes doesn't contain an x velocity and a y velocity.
+    """
+    from iris import Constraint
+    from iris.cube import CubeList
+    import numpy as np
+
```
Comment (benfitzpatrick) [inline lib/improver/cli/nowcast_accumulate.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948)

Comment (neilCrosswaite) [inline lib/improver/cli/nowcast_accumulate.py]:
```suggestion

```

File: lib/improver/cli/nowcast_accumulate.py (hunk 8 above)
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_accumulate.py]: Done

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

File: lib/improver/cli/nowcast_accumulate.py [hunk 15]
```diff
@@ -101,13 +106,10 @@ def process(input_cube: cli.inputcube,
             "Neither u_cube or v_cube can be none")
```
Comment (arh89) [inline lib/improver/cli/nowcast_accumulate.py]:
```Suggestion
            "Neither u_cube or v_cube can be None")
```

File: lib/improver/cli/nowcast_accumulate.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+    u_cube = advection_cubes.extract(
+        Constraint("precipitation_advection_x_velocity"), True)
+    v_cube = advection_cubes.extract(
+        Constraint("precipitation_advection_y_velocity"), True)
+
+    if not (u_cube and v_cube):
+        raise TypeError(
+            "Neither u_cube or v_cube can be none")
+    oe_cube = merge_cubes(CubeList(oe_cube))
```
Comment (arh89) [inline lib/improver/cli/nowcast_accumulate.py]: I don't think we need this?

File: lib/improver/cli/nowcast_accumulate.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+
+inputadvection = cli.create_constrained_inputcubelist_converter(
+    'precipitation_advection_x_velocity', 'precipitation_advection_y_velocity')
+
+
+@cli.clizefy
+@cli.with_output
+def process(input_cube: cli.inputcube,
+            advection_cubes: inputadvection,
+            *oe_cube: cli.inputcube,
```
Comment (arh89) [inline lib/improver/cli/nowcast_accumulate.py]:
```Suggestion
oe_cube: cli.inputcube,
*,

File: lib/improver/cli/nowcast_accumulate.py (hunk 16 above)
Comment (arh89) [reply lib/improver/cli/nowcast_accumulate.py]:
```suggestion
```

File: tests/improver-nowcast-accumulate/00-basic.bats [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "accumulations basics" {
+  improver_check_skip_acceptance
+  KGO="nowcast-accumulate/basic/kgo.nc"
+
+  UVCOMP="$IMPROVER_ACC_TEST_DIR/nowcast-optical-flow/basic/kgo.nc"
+  INFILE="201811031600_radar_rainrate_composite_UK_regridded.nc"
+  OE1="20181103T1600Z-PT0003H00M-orographic_enhancement.nc"
```
Comment (bayliffe) [inline tests/improver-nowcast-accumulate/00-basic.bats]: Why have you defined one as a complete path and two with partial paths? Simply to differentiate those coming from a different CLI's test data?

File: lib/improver/cli/__init__.py [hunk 19]
```diff
@@ -173,6 +173,36 @@ def comma_separated_list(to_convert):
...
+        *constraints (str):
+            constraints to be used in the loading of cubes against a cubeList
+
+    Returns:
+        list of function:
+            A list of functions with different constraints.
+    """
+    @value_converter
+    def constrained_inputcubelist_converter(to_convert):
+        """Passes the cube and constraints onto maybe coerce with.
```
[the remark this answers is not in the document]
Comment (arh89) [reply lib/improver/cli/__init__.py]: `inputcube` is used elsewhere as another converter, so the placement of the underscores is intended to indicate the relationship between the two of them.

File: tests/improver-nowcast-accumulate/00-basic.bats (hunk 18 above)
Comment (neilCrosswaite) [reply tests/improver-nowcast-accumulate/00-basic.bats]: Yes, I believe that i really does lend itself to being more unstandable to a new reader.

File: lib/improver/cli/__init__.py [hunk 20]
```diff
@@ -173,6 +173,36 @@ def comma_separated_list(to_convert):
...
+
+def create_constrained_inputcubelist_converter(*constraints):
+    """Makes a list of functions with different constraints.
+
+    Args:
+        *constraints (str):
+            constraints to be used in the loading of cubes against a cubeList
+
+    Returns:
+        list of function:
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: For the ret of the return types we use `list of str` or `list of int`, It reads weird but the `type()` of a function is function.

File: lib/improver/cli/__init__.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/cli/__init__.py]: Yeah, I meant in the doc string where it suddenly has the weird sentence "maybe coerce with".

File: tests/improver-nowcast-accumulate/00-basic.bats (hunk 18 above)
Comment (bayliffe) [reply tests/improver-nowcast-accumulate/00-basic.bats]: 😆

File: lib/improver/cli/__init__.py [hunk 21]
```diff
@@ -173,6 +173,36 @@ def comma_separated_list(to_convert):
...
+            A list of functions with different constraints.
+    """
+    @value_converter
+    def constrained_inputcubelist_converter(to_convert):
+        """Passes the cube and constraints onto maybe coerce with.
+
+        Args (string or iris.cube.CubeList):
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: Fixed.

File: lib/improver/cli/__init__.py [hunk 22]
```diff
@@ -173,6 +173,36 @@ def comma_separated_list(to_convert):
...
+            To cube to be passed forward for returning or loading.
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: Fixed

File: lib/improver/cli/__init__.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: But also, it just returns a function and not a list of them.

File: lib/improver/tests/cli/test_init.py [hunk 23]
```diff
@@ -185,6 +186,20 @@ def test_with_output(self, m):
         self.assertEqual(result, 4)
 
 
+class Test_create_constrained_inputcube_converter(unittest.TestCase):
+    """Tests the creature constraint inputcube converter"""
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/tests/cli/test_init.py]: See earlier.

File: lib/improver/cli/__init__.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (arh89) [reply lib/improver/cli/__init__.py]: Ah, gotcha, my bad!

File: lib/improver/tests/cli/test_init.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (arh89) [reply lib/improver/tests/cli/test_init.py]:
```suggestion
    """Tests the create_constrained_inputcube_converter"""
```

File: lib/improver/cli/__init__.py [hunk 24]
```diff
@@ -173,6 +173,36 @@ def comma_separated_list(to_convert):
...
+            iris.cube.CubeList:
+                The loaded cubelist of constrained cubes.
+        """
+        from improver.utilities.load import load_cube
+        from iris.cube import CubeList
+        return CubeList([maybe_coerce_with(
+            load_cube, to_convert, constraints=j) for j in constraints])
+
+    return constrained_inputcubelist_converter
+
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: New docstring
```    
    The function is value_converter, this means it is used by clize to convert
    strings into objects.
    This is a way of not using the IMPROVER load_cube which will try to merge
    cubes. Iris load on the other hand won't deal with meta data properly.
    So an example if you wanted to load an X cube and a Y cube from a cubelist
    of 2. You call this function with a list of constraints.
    These cubes get loaded and returned as a CubeList.
```

File: lib/improver/cli/nowcast_accumulate.py [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+    u_cube = advection_cubes.extract(
+        Constraint("precipitation_advection_x_velocity"), True)
+    v_cube = advection_cubes.extract(
+        Constraint("precipitation_advection_y_velocity"), True)
+
+    if not (u_cube and v_cube):
+        raise TypeError(
+             "Neither u_cube or v_cube can be None")
+
+    # extrapolate input data to required lead times
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_accumulate.py]: typo?

File: lib/improver/cli/nowcast_accumulate.py [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+    forecast_cubes = CreateExtrapolationForecast(
+        input_cube, u_cube, v_cube, orographic_enhancement_cube=oe_cube,
+        attributes_dict=attributes_dict).process(ACCUMULATION_FIDELITY,
+                                                 max_lead_time)
+
+    lead_times = (np.arange(lead_time_interval, max_lead_time + 1,
+                            lead_time_interval))
+
```
[the remark this answers is not in the document]
Comment (arh89) [reply lib/improver/cli/nowcast_accumulate.py]:
```suggestion
    # Accumulate high frequency rate extrapolations into desired accumulation intervals.
```

File: lib/improver/cli/nowcast_accumulate.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (arh89) [reply lib/improver/cli/nowcast_accumulate.py]:
```suggestion
    # extrapolate input data to maximum required lead times
```

File: lib/improver/cli/nowcast_accumulate.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_accumulate.py]: Due to line length constraints.
`# Accumulate high frequency rate into desired accumulation intervals.`

File: lib/improver/cli/nowcast_accumulate.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/nowcast_accumulate.py]:
```suggestion
    # extrapolate input data to maximum required lead time
```

File: lib/improver/cli/__init__.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/__init__.py]: Done

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

File: lib/improver/cli/nowcast_accumulate.py [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+    Returns:
+        iris.cube.CubeList:
+            New cubes with accumulated data.
+
+    Raises:
+        TypeError:
+            If advection_cubes doesn't contain an x velocity and a y velocity.
+    """
+    from iris import Constraint
+    from iris.cube import CubeList
```
Comment (benfitzpatrick) [inline lib/improver/cli/nowcast_accumulate.py]: Codacy Issue found: [Unused CubeList imported from iris.cube](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

File: lib/improver/cli/nowcast_accumulate.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (arh89) [reply lib/improver/cli/nowcast_accumulate.py]:
```suggestion
    # extrapolate input data to the maximum required lead time
```
...just to carry on the suggestion-train

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- lib/improver/cli/nowcast_accumulate.py  3
- lib/improver/tests/acceptance/test_nowcast_accumulate.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

File: lib/improver/tests/acceptance/test_nowcast_accumulate.py [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+    output_path = tmp_path / "output.nc"
+
+    args = [input_path, uv_path, oe_path,
+            "--max-lead-time", "30",
+            "--output", output_path]
+
+    run_cli(args)
+    acc.compare(output_path, kgo_path)
+
+
```
Comment (benfitzpatrick) [inline lib/improver/tests/acceptance/test_nowcast_accumulate.py]: Codacy Issue found: [Trailing newlines](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- lib/improver/tests/acceptance/test_nowcast_accumulate.py  1
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)

File: lib/improver/tests/acceptance/test_nowcast_accumulate.py (hunk 28 above)
Comment (benfitzpatrick) [inline lib/improver/tests/acceptance/test_nowcast_accumulate.py]: Codacy Issue found: [Trailing newlines](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/tests/acceptance/test_nowcast_accumulate.py  1
- lib/improver/cli/nowcast_accumulate.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4628948&bid=15516462)
